frame/bags-list/src/list/mod.rs

```diff
@@ -0,0 +1,670 @@
+	}
+
+	/// Migrate the list from one set of thresholds to another.
+	///
+	/// This should only be called as part of an intentional migration; it's fairly expensive.
+	///
+	/// Returns the number of accounts affected.
+	///
+	/// Preconditions:
+	///
+	/// - `old_thresholds` is the previous list of thresholds.
+	/// - All `bag_upper` currently in storage are members of `old_thresholds`.
+	/// - `T::BagThresholds` has already been updated and is the new set of thresholds.
+	///
+	/// Postconditions:
+	///
+	/// - All `bag_upper` currently in storage are members of `T::BagThresholds`.
+	/// - No id is changed unless required to by the difference between the old threshold list and
+	///   the new.
+	/// - ids whose bags change at all are implicitly rebagged into the appropriate bag in the new
+	///   threshold set.
+	#[allow(dead_code)]
+	pub fn migrate(old_thresholds: &[VoteWeight]) -> u32 {
+		// we can't check all preconditions, but we can check one
+		debug_assert!(
+			crate::ListBags::<T>::iter().all(|(threshold, _)| old_thresholds.contains(&threshold)),
+			"not all `bag_upper` currently in storage are members of `old_thresholds`",
+		);
+
+		let old_set: BTreeSet<_> = old_thresholds.iter().copied().collect();
+		let new_set: BTreeSet<_> = T::BagThresholds::get().iter().copied().collect();
+
+		// accounts that need to be rebagged
+		let mut affected_accounts = BTreeSet::new();
+		// track affected old bags to make sure we only iterate them once
+		let mut affected_old_bags = BTreeSet::new();
+
+		let new_bags = new_set.difference(&old_set).copied();
+		// a new bag means that all accounts previously using the old bag's threshold must now
+		// be rebagged
+		for inserted_bag in new_bags {
+			let affected_bag = {
+				// this recreates `notional_bag_for` logic, but with the old thresholds.
+				let idx = old_thresholds.partition_point(|&threshold| inserted_bag > threshold);
+				old_thresholds.get(idx).copied().unwrap_or(VoteWeight::MAX)
+			};
+			if !affected_old_bags.insert(affected_bag) {
+				// If the previous threshold list was [10, 20], and we insert [3, 5], then there's
+				// no point iterating through bag 10 twice.
+				continue
+			}
+
+			if let Some(bag) = Bag::<T>::get(affected_bag) {
+				affected_accounts.extend(bag.iter().map(|node| node.id));
+			}
+		}
+
+		let removed_bags = old_set.difference(&new_set).copied();
+		// a removed bag means that all members of that bag must be rebagged
+		for removed_bag in removed_bags.clone() {
+			if !affected_old_bags.insert(removed_bag) {
+				continue
+			}
+
+			if let Some(bag) = Bag::<T>::get(removed_bag) {
+				affected_accounts.extend(bag.iter().map(|node| node.id));
+			}
+		}
+
+		// migrate the voters whose bag has changed
+		let weight_of = T::VoteWeightProvider::vote_weight;
+		Self::remove_many(&affected_accounts);
+		let num_affected = affected_accounts.len() as u32;
+		let _ = Self::insert_many(affected_accounts.into_iter(), weight_of);
```

Comment:
```suggestion
		let _inserted = Self::insert_many(affected_accounts.into_iter(), weight_of);
		debug_assert_eq!(_inserted, num_affected);
```

Reply:
597e1865a7f8c509b56977428a60be8a9b2d4df1 we now also checked the number removed